app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,66 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
+  map: Map<string, WeakRefCompatibleObject<Record<string, object>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
+    Object.keys(value).forEach((valueKey) => {
+      weakReffedValueObj[valueKey] = new WeakRef(value[valueKey]);
+    });
+    this.map.set(key, weakReffedValueObj);
+  }
+
+  get(key: string) {
+    const weakReffedValue = this.map.get(key);
+    if (!weakReffedValue) {
+      return undefined;
+    }
+
+    const deReffedValue = {} as { [dereffedValueObjKey: string]: WeakKey };
+    for (const weakReffedValueKey of Object.keys(weakReffedValue)) {
+      const deref = weakReffedValue[weakReffedValueKey].deref();
+      if (deref === undefined) {
+        this.map.delete(weakReffedValueKey);
+        return undefined;
+      }
+      deReffedValue[weakReffedValueKey] = deref;
+    }
+
+    return deReffedValue;
+  }
+
+  has(key: string) {
+    return this.get(key) !== undefined;
+  }
+
+  delete(key: string) {
+    const value = this.get(key);
+    if (value !== undefined) {
+      return this.map.delete(key);
+    }
+    return false;
+  }
+
+  clear() {
+    this.map.clear();
+  }
+
+  forEach(callback: (value: WeakRefCompatibleObject<Record<string, object>>, key: string, map: Map<string, WeakRefCompatibleObject<Record<string, object>>>) => void, thisArg?: any): void {
```

Comment:
Adding this pass through for forEach to access the internal map for this custom Map wrapper [so that it can be accessed cleanly here](https://github.com/MetaMask/core/pull/4130/files#r1558425789) and preserve the Map typing for the `#domainProxyMap` variable on the `SelectedNetworkController` where its used...